### DISCUSSION TITLE
How to test if a connection is still valid

### DISCUSSION THREAD

**Discussion by Sell24** [reactphp/reactphp#542] [Q&A] [answered]
I have a ReactPHP Server set up as a socket service in Laravel which listens for incoming connections and then receives data from a device. 

Unfortunately, this embedded device is not capable of a graceful shutdown so it does not close the connection when it is turned off. So when I look the ports on the firewall, it still shows them as open, thus not releasing those resources.

I monitor both the 'end' and 'close' events on the server side, but neither of these events seems to be triggered when the device stops transmitting.  So I just wanted to check, am I supposed to get an event if the connection is closed abruptly by the client or will I need to implement some sort of timeout?

TIA

**Comment by SimonFrings** [comment, marked as answer]
Hey @Sell24, thanks for bringing this up :+1:

In a gist: If the device you're referring to is not capable of a graceful shutdown, it will also not send an `end` or `close` event. This means that if the device "dies" for any reason, it will be unable to notify you that the connection is "dead". To determine if your communication partner is still alive, you can implement higher-level application timeout logic, such as sending a heartbeat message every once in a while. If the device is dead and does not respond to your heartbeat message, you can close the connection.

There are similar tickets also covering this aspect like https://github.com/reactphp/socket/issues/193, https://github.com/reactphp/socket/issues/249 and others.

Hope this helps :)

**Reply by Sell24** [reply]
Perfect. Thanks for the quick reply. I missed reactphp/socket#193 when I was looking at the issues. Sorry about that but thanks again.
